### PR TITLE
Only check for <link> elements if response is HTML

### PR DESCRIPTION
**Pull request by jrf0110**
What this PR solves / how to test:

This prevents the Pages Asset Server from performing unnecessary checks for preload/preconnect links. I've reproduced the unintended behavior and wrote an internal integration test to ensure that this change solves the problem.

Associated docs issues/PR:

- N/A

Author has included the following, where applicable:

- [ ] Tests
- [ ] Changeset

Reviewer has performed the following, where applicable:

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [x] Manually pulled down the changes and spot-tested

### CODE REVIEW

File: packages/pages-shared/asset-server/handler.ts [hunk 1]
```diff
@@ -308,12 +308,14 @@ export async function generateHandler<
 
 			const clonedResponse = response.clone();
 
-			if (waitUntil) {
+			if (
+				waitUntil &&
+				response.headers.get("Content-Type")?.includes("text/html")
```
Comment (GregBrimble) [inline packages/pages-shared/asset-server/handler.ts:313]: Further down, we do `contentType.startsWith("text/html")`: https://github.com/cloudflare/wrangler2/blob/cbbc932f931c45b51ed735557e7b7ded9fbf1c34/packages/pages-shared/asset-server/handler.ts#L509

Should that be `contentType.includes("text/html")` like you have here? Maybe worth just extracting out to an `isHTML: (contentType: string) => boolean` so we're consistent in both places?

**Comment by GregBrimble** [comment]
You'll need a changeset to make this release in a new `@cloudflare/pages-shared` version.

File: packages/pages-shared/asset-server/handler.ts (hunk 1 above)
Comment (jrf0110) [reply packages/pages-shared/asset-server/handler.ts:322]: Actually, I think it _should be_ `.startsWith(...)`. At the time of me writing the code, I wasn't sure if the type could be anywhere in the string, but it does seem that clients expect the type to come first, then directives. I'll abstract to a function and use `.startsWith(...)`

File: packages/pages-shared/__tests__/jest.setup.ts [hunk 2]
```diff
@@ -1,3 +1,10 @@
-import makeServiceWorkerEnv from "service-worker-mock";
+import { fetch, Headers, Request, Response } from "@miniflare/core";
+import { HTMLRewriter } from "@miniflare/html-rewriter";
 
-Object.assign(globalThis, makeServiceWorkerEnv());
```
Comment (jrf0110) [inline packages/pages-shared/__tests__/jest.setup.ts:3]: I was getting issues with service-worker-mock not supporting `response.arrayBuffer()` which miniflare's in-memory cache implementation relies on. So I got rid of it!

Comment (jrf0110) [reply packages/pages-shared/__tests__/jest.setup.ts:3]: I should add a comment here explaining that we need to provide API's to the global scope as needed

File: packages/pages-shared/asset-server/handler.ts [hunk 3]
```diff
@@ -279,11 +279,12 @@ export async function generateHandler<
...
 			"referrer-policy": "strict-origin-when-cross-origin",
-			...(existingHeaders.has("content-type")
-				? { "x-content-type-options": "nosniff" }
-				: {}),
 		});
 
+		if (existingHeaders.has("content-type")) {
+			extraHeaders.append("x-content-type-options", "nosniff");
+		}
+
```
Comment (jrf0110) [inline packages/pages-shared/asset-server/handler.ts:287]: I should revert this change

File: packages/pages-shared/asset-server/handler.ts [hunk 4]
```diff
@@ -296,6 +297,7 @@ export async function generateHandler<
 			// "Early Hints cache entries are keyed by request URI and ignore query strings."
 			// https://developers.cloudflare.com/cache/about/early-hints/
 			const earlyHintsCacheKey = `${protocol}//${host}${pathname}`;
+			console.log("earlyHintsCacheKey", earlyHintsCacheKey);
```
Comment (jrf0110) [inline packages/pages-shared/asset-server/handler.ts:300]: Need to remove these debug logs

File: packages/pages-shared/__tests__/jest.setup.ts (hunk 2 above)
Comment (GregBrimble) [reply packages/pages-shared/__tests__/jest.setup.ts:3]: Maybe add a `miniflare-test` polyfill and use that here?


https://github.com/cloudflare/wrangler2/blob/main/packages/pages-shared/environment-polyfills/miniflare.ts

File: .changeset/shaggy-hornets-drop.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@cloudflare/pages-shared": patch
+---
+
+Only check for <link> elements if response is HTML in Pages asset handler early hints code path.
```
Comment (GregBrimble) [inline .changeset/shaggy-hornets-drop.md:5]:
```suggestion
Only check for `<link>` elements if the asset response is HTML in Pages' Early Hints feature.
```

File: packages/pages-shared/__tests__/asset-server/handler.test.ts [hunk 6]
```diff
@@ -391,6 +500,18 @@ async function getTestResponse({
 				}
 			);
 		},
+		waitUntil: (async (promise: Promise<unknown>) => {
+			spies.waitUntil.push(promise);
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		}) as any,
```
Comment (GregBrimble) [inline packages/pages-shared/__tests__/asset-server/handler.test.ts:506]:
```suggestion
		waitUntil: async (promise: Promise<unknown>) => {
			spies.waitUntil.push(promise);
		},
```

Don't think you need that `any`? My editor isn't complaining, at least :)

File: packages/pages-shared/asset-server/handler.ts [hunk 7]
```diff
@@ -352,7 +355,12 @@ export async function generateHandler<
 							if (linkHeader) {
 								await earlyHintsCache.put(
 									earlyHintsCacheKey,
-									new Response(null, { headers: { Link: linkHeader } })
+									new Response(null, {
+										headers: {
+											Link: linkHeader,
+											"Cache-Control": "max-age=86400",
```
[the remark this answers is not in the document]
Comment (GregBrimble) [reply packages/pages-shared/asset-server/handler.ts:360]: This isn't 30 days. It's 1 day. I'd support anything 2hr+, so whatever — just wanted to flag to you.

File: packages/pages-shared/environment-polyfills/miniflare.ts [hunk 8]
```diff
@@ -2,11 +2,13 @@ import { polyfill } from ".";
 
 export default async () => {
 	const mf = await import("@miniflare/core");
+	const { HTMLRewriter } = await import("@miniflare/html-rewriter");
```
Comment (GregBrimble) [inline packages/pages-shared/environment-polyfills/miniflare.ts:5]: We aren't passing `caches`/`waitUntil` to the `generateHandler()` when we actually use this in `wrangler pages dev` (https://github.com/cloudflare/wrangler2/blob/main/packages/wrangler/src/miniflare-cli/assets.ts#L170-L237) so why have you added `HTMLRewriter` here?

File: packages/pages-shared/asset-server/handler.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jrf0110) [reply packages/pages-shared/asset-server/handler.ts:360]: How to math with John Fawcett 💯

File: packages/pages-shared/__tests__/asset-server/handler.test.ts (hunk 6 above)
Comment (jrf0110) [reply packages/pages-shared/__tests__/asset-server/handler.test.ts:506]: Ah thank you. I was moving fast there and meant to come back to that

File: packages/pages-shared/environment-polyfills/miniflare.ts (hunk 8 above)
Comment (jrf0110) [reply packages/pages-shared/environment-polyfills/miniflare.ts:5]: I should revert this. I was originally trying to use this function directly in the jest setup file.

Comment (jrf0110) [reply packages/pages-shared/environment-polyfills/miniflare.ts:5]: In [packages/pages-shared/environment-polyfills/types.ts](https://github.com/cloudflare/wrangler2/blob/main/packages/pages-shared/environment-polyfills/types.ts#L11) we declare it as a global that's available and we use it freely. The test code passes in `caches` and `waitUntil` so they can be spied upon

Comment (GregBrimble) [reply packages/pages-shared/environment-polyfills/miniflare.ts:5]: Ah, fair enough then. I'd be tempted to pull it out of global and pass it through like the rest of these things — I must have just overlooked it before — but no biggie and adding it here is fine :)

**Review by GregBrimble — APPROVED** [review]
Approved for when you get your tests passing. Gimme a shout if you need a hand.
